compiler/noirc_frontend/src/parser/parser.rs

```diff
@@ -91,1764 +44,504 @@ use traits::where_clause;
+        Ok(item)
+    } else {
+        Err(parser.errors)
+    }
 }
 
-fn self_parameter() -> impl NoirParser<Param> {
-    let mut_ref_pattern = just(Token::Ampersand).then_ignore(keyword(Keyword::Mut));
-    let mut_pattern = keyword(Keyword::Mut);
-
-    mut_ref_pattern
-        .or(mut_pattern)
-        .map_with_span(|token, span| (token, span))
-        .or_not()
-        .then(filter_map(move |span, found: Token| match found {
-            Token::Ident(ref word) if word == "self" => Ok(span),
-            _ => Err(ParserError::expected_label(ParsingRuleLabel::Parameter, found, span)),
-        }))
-        .map(|(pattern_keyword, ident_span)| {
-            let ident = Ident::new("self".to_string(), ident_span);
-            let path = Path::from_single("Self".to_owned(), ident_span);
-            let no_args = GenericTypeArgs::default();
-            let mut self_type =
-                UnresolvedTypeData::Named(path, no_args, true).with_span(ident_span);
-            let mut pattern = Pattern::Identifier(ident);
-
-            match pattern_keyword {
-                Some((Token::Ampersand, _)) => {
-                    self_type = UnresolvedTypeData::MutableReference(Box::new(self_type))
-                        .with_span(ident_span);
-                }
-                Some((Token::Keyword(_), span)) => {
-                    pattern = Pattern::Mutable(Box::new(pattern), span.merge(ident_span), true);
-                }
-                _ => (),
-            }
-
-            Param { span: pattern.span(), pattern, typ: self_type, visibility: Visibility::Private }
-        })
+enum TokenStream<'a> {
+    Lexer(Lexer<'a>),
+    Tokens(Tokens),
 }
 
-/// Function declaration parameters differ from other parameters in that parameter
-/// patterns are not allowed in declarations. All parameters must be identifiers.
-fn function_declaration_parameters() -> impl NoirParser<Vec<(Ident, UnresolvedType)>> {
-    let typ = parse_type().recover_via(parameter_recovery());
-    let typ = just(Token::Colon).ignore_then(typ);
-
-    let full_parameter = ident().recover_via(parameter_name_recovery()).then(typ);
-    let self_parameter = self_parameter().validate(|param, span, emit| {
-        match param.pattern {
-            Pattern::Identifier(ident) => (ident, param.typ),
-            other => {
-                emit(ParserError::with_reason(
-                    ParserErrorReason::PatternInTraitFunctionParameter,
-                    span,
-                ));
-                // into_ident panics on tuple or struct patterns but should be fine to call here
-                // since the `self` parser can only parse `self`, `mut self` or `&mut self`.
-                (other.into_ident(), param.typ)
-            }
+impl<'a> TokenStream<'a> {
+    fn next(&mut self) -> Option<SpannedTokenResult> {
+        match self {
+            TokenStream::Lexer(lexer) => lexer.next(),
+            TokenStream::Tokens(tokens) => tokens.0.pop().map(Ok),
```

Comment:
Wouldn't this pop in reverse order?

Edit: Ah, I see we `.reverse()` the tokens in `for_tokens`. Can you add a comment here that that is an invariant?